### PR TITLE
xds: Add ext_authz response handling

### PR DESCRIPTION
**Pull request by sauravzg**
Part 2 of the client-side ext_authz filter. Sits on top of #12493.

Adds CheckResponseHandler, which interprets the CheckResponse from the authorization service. It evaluates OkHttpResponse vs DeniedHttpResponse, maps HTTP status codes to gRPC statuses, applies failure_mode_allow semantics when the authz server is unreachable, and validates decoder header mutations against the configured HeaderMutationRulesConfig.

AuthzResponse is the resulting value object carrying the allow/deny decision, the gRPC status for denied calls, and any header/trailer mutations to apply.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Handles the response from the external authorization service, processing it to determine the
+ * authorization decision and applying any necessary header mutations.
+ */
+@ThreadSafe
+public class CheckResponseHandler {
+  private final HeaderMutationFilter headerMutationFilter;
+
+  public CheckResponseHandler(HeaderMutationFilter headerMutationFilter) {
```
Comment (kannanjgithub) [inline xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java:42]: Make the class and the constructor package-private.

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    }
+  }
+
+  private AuthzResponse handleOkResponse(final CheckResponse response)
+      throws HeaderMutationDisallowedException {
+    if (!response.hasOkResponse()) {
+      return AuthzResponse.allow(
+          HeaderMutations.create(ImmutableList.of(), ImmutableList.of())).build();
+    }
+    OkHttpResponse okResponse = response.getOkResponse();
```
Comment (kannanjgithub) [inline xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java:64]: nit: Can just inline the expression without using a local variable.

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    CheckResponseMutations allowedMutations = buildHeaderMutationsFromOkResponse(okResponse);
+
+    return AuthzResponse.allow(allowedMutations.requestMutations())
+        .setResponseHeaderMutations(allowedMutations.responseMutations()).build();
```
Comment (kannanjgithub) [inline xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java:68]: Looking at this usage, the response mutations can also be an argument to `AuthzResponse.allow` instead of making it set EMPTY_MUTATIONS for response.

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    String baseMsg = "RPC denied by external authorization server";
+    String outerMsg = response.getStatus().getMessage();
+    String description = outerMsg.isEmpty() ? baseMsg : baseMsg + ": " + outerMsg;
+
+    if (!response.hasDeniedResponse()) {
+      return AuthzResponse.deny(Status.PERMISSION_DENIED.withDescription(description)).build();
+    }
+    DeniedHttpResponse deniedResponse = response.getDeniedResponse();
+    CheckResponseMutations allowedMutations =
+        buildHeaderMutationsFromDeniedResponse(deniedResponse);
```
Comment (kannanjgithub) [inline xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java:95]: Since only response mutations is used for the denied case, can we either avoid calling this method that creates mutations for both request and response, or add a boolean argument to indicate skipping request mutations?

File: xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+    AuthzResponse authzResponse = responseHandler.handleResponse(checkResponse);
+
+    assertThat(authzResponse.decision()).isEqualTo(Decision.DENY);
+    assertThat(authzResponse.status().get().getCode()).isEqualTo(Status.INTERNAL.getCode());
+    assertThat(authzResponse.status().get().getDescription()).isEqualTo("disallowed");
+  }
+
+  @Test
+  public void handleResponse_ok_edgeCaseHeaders() {
```
Comment (kannanjgithub) [inline xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java:178]: nit: Can we use a more indicative name that tells what edge case it is?

File: xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        .setStatus(com.google.rpc.Status.newBuilder().setCode(Code.OK_VALUE).build()).build();
+    AuthzResponse authzResponse = responseHandler.handleResponse(checkResponse);
+    assertThat(authzResponse.decision()).isEqualTo(Decision.ALLOW);
+    assertThat(authzResponse.requestHeaderMutations())
+        .isEqualTo(HeaderMutations.create(ImmutableList.of(), ImmutableList.of()));
+  }
+
+  @Test
+  public void handleResponse_okWithMutations() {
+    HeaderValueOption option =
```
Comment (kannanjgithub) [inline xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java:79]: Can we also add envoy HeaderValueOptions for ADD_IF_ABSENT, OVERWRITE_IF_EXISTS_OR_ADD and OVERWRITE_IF_EXISTS and verify their translation in the same test?

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java (hunk 1 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java]: Done.

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java (hunk 2 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java]: Done.

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java (hunk 3 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java]: Done. Added an overload for Build

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java (hunk 4 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java]: Done.

File: xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java (hunk 5 above)
Comment (sauravzg) [reply xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java]: Done.

File: xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java (hunk 6 above)
Comment (sauravzg) [reply xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java]: Done.

File: xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+            .newBuilder().setKey("test-key-3").setValue("val-3"))
+            .setAppendAction(HeaderValueOption.HeaderAppendAction.OVERWRITE_IF_EXISTS_OR_ADD)
+            .build();
+    HeaderValueOption option4 =
+        HeaderValueOption.newBuilder().setHeader(HeaderValue
+            .newBuilder().setKey("test-key-4").setValue("val-4"))
+            .setAppendAction(HeaderValueOption.HeaderAppendAction.OVERWRITE_IF_EXISTS)
+            .build();
+
+    io.grpc.xds.internal.headermutations.HeaderValueOption expectedOption1 =
```
Comment (kannanjgithub) [inline xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java:100]: nit: Use the imported class.

**Review by kannanjgithub — APPROVED** [review]

File: xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java (hunk 7 above)
Comment (sauravzg) [reply xds/src/test/java/io/grpc/xds/internal/extauthz/CheckResponseHandlerTest.java:100]: This is WAI. The imported `HeaderValueOption` at the top of the file is Envoy's protobuf message (`io.envoyproxy.envoy.config.core.v3.HeaderValueOption`). Since our internal domain class has the identical simple name (`io.grpc.xds.internal.headermutations.HeaderValueOption`) and Java does not support import aliasing, the internal class is intentionally fully qualified here to avoid collision.

File: xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java (hunk 1 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/internal/extauthz/CheckResponseHandler.java]: Noting that I had to revert this change, because I need this to be public as it's used in #12898 . Making this package private and constructing it internally could result in a performance tradeoff where we construct every rpc instead of every interceptor build.
